### PR TITLE
gh-104252: Immortalize Py_EMPTY_KEYS

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
This was missed in gh-19474.  It matters for with a per-interpreter GIL since `PyDictKeysObject.dk_refcnt` breaks isolation and leads to races.

cc @eduardo-elizondo

* Issue: gh-104252

### CODE REVIEW

File: Objects/dictobject.c [hunk 1]
```diff
@@ -300,18 +300,26 @@ _PyDict_DebugMallocStats(FILE *out)
 
 static void free_keys_object(PyInterpreterState *interp, PyDictKeysObject *keys);
 
+// XXX Switch to Py_INCREF()?
```
Comment (sunmy2019) [inline Objects/dictobject.c]: what does this line mean?
`PyDictKeysObject` has a different layout from `PyObject`.

**Review by sunmy2019 — CHANGES_REQUESTED** [review]
There are other places with direct changes to `dk_refcnt`. They are not guarded.

https://github.com/python/cpython/blob/472938316a85c706c06ad1b3727a205d5bffcb1f/Objects/dictobject.c#L639

https://github.com/python/cpython/blob/472938316a85c706c06ad1b3727a205d5bffcb1f/Objects/dictobject.c#L822-L828

https://github.com/python/cpython/blob/472938316a85c706c06ad1b3727a205d5bffcb1f/Objects/dictobject.c#L1530-L1538

File: Objects/dictobject.c (hunk 1 above)
Comment (ericsnowcurrently) [reply Objects/dictobject.c]: Right, any switch would require a slight change to the `PyDictObject` layout, but only if `Py_TRACE_REFS` is defined.  `PyObject.ob_refcnt` is otherwise at the beginning of the struct, so the rest wouldn't matter.

That said, the only reason why it would matter is to get the performance benefits of `Py_INCREF()` for immortal objects.  In the case of `Py_EMPTY_KEYS` I expect the benefit would be negligible, so it mostly wouldn't be worth the trouble.

Normally I'd clarify the comment, but I'll probably drop it instead.

**Comment by ericsnowcurrently** [comment]
@sunmy2019, thanks for pointing out the other uses of `dk_refcnt`.  I'll fix that.

File: Objects/dictobject.c (hunk 1 above)
Comment (ericsnowcurrently) [reply Objects/dictobject.c]: done

Comment (ericsnowcurrently) [reply Objects/dictobject.c]: done

File: Objects/dictobject.c [hunk 2]
```diff
@@ -3565,7 +3574,7 @@ _PyDict_SizeOf(PyDictObject *mp)
     }
     /* If the dictionary is split, the keys portion is accounted-for
        in the type object. */
-    if (mp->ma_keys->dk_refcnt == 1) {
+    if (mp->ma_keys->dk_refcnt == 1 || mp->ma_keys == Py_EMPTY_KEYS) {
```
Comment (methane) [inline Objects/dictobject.c]: Why is this change needed?

Comment (ericsnowcurrently) [reply Objects/dictobject.c]: Hmm, I suppose it isn't, since an empty ma_keys has a size of 0.  I'll drop that change.

File: Include/internal/pycore_dict_state.h [hunk 3]
```diff
@@ -27,6 +27,8 @@ struct _Py_dict_state {
     uint64_t global_version;
     uint32_t next_keys_version;
 
+//    PyDictKeysObject empty_keys_struct;
```
Comment (sunmy2019) [inline Include/internal/pycore_dict_state.h]: forget to remove?

Comment (ericsnowcurrently) [reply Include/internal/pycore_dict_state.h]: fixed

File: Objects/dictobject.c (hunk 2 above)
Comment (ericsnowcurrently) [reply Objects/dictobject.c]: fixed

File: Objects/dictobject.c [hunk 4]
```diff
@@ -300,9 +300,18 @@ _PyDict_DebugMallocStats(FILE *out)
 
 static void free_keys_object(PyInterpreterState *interp, PyDictKeysObject *keys);
 
+/* We might consider modifying PyDictKeysObject so we could use
+   Py_INCREF() instead of dictkeys_incref() (and likewise with
+   Py_DECREF() and dictkeys_decref()).  However, there doesn't
+   seem to be enough benefit (performance or otherwise) to justify
+   the change. */
```
Comment (methane) [inline Objects/dictobject.c]: I don't like this comment. Since dictkeys doesn't have type pointer, it is too trickey to use `Py_DECREF()`.

File: Objects/dictobject.c [hunk 5]
```diff
@@ -2080,8 +2089,8 @@ PyDict_Clear(PyObject *op)
         dictkeys_decref(interp, oldkeys);
     }
     else {
-       assert(oldkeys->dk_refcnt == 1);
-       dictkeys_decref(interp, oldkeys);
+        assert(oldkeys->dk_refcnt == 1 || oldkeys == Py_EMPTY_KEYS);
+        dictkeys_decref(interp, oldkeys);
```
Comment (methane) [inline Objects/dictobject.c:2094]: Nice catch about wrong indent.
I think no need to update the assert, because we have this code above:

```c
    if (oldkeys == Py_EMPTY_KEYS) {
        return;
    }
```

**Comment by methane** [comment]
Please remove these lines too:

https://github.com/python/cpython/blob/58585a9a983002c2d76c9524d60750d32c7d69ef/Objects/dictobject.c#L849

https://github.com/python/cpython/blob/58585a9a983002c2d76c9524d60750d32c7d69ef/Objects/dictobject.c#L1347

And Move `_Py_DecRefTotal()` into `if (oldkeys != Py_EMPTY_KEYS)` block.

https://github.com/python/cpython/blob/58585a9a983002c2d76c9524d60750d32c7d69ef/Objects/dictobject.c#L1545-L1548

**Comment by ericsnowcurrently** [comment]
Thanks for the feedback.  I'll clean things up tomorrow.

File: Objects/dictobject.c (hunk 4 above)
Comment (ericsnowcurrently) [reply Objects/dictobject.c]: I'll fix the comment.

Comment (ericsnowcurrently) [reply Objects/dictobject.c]: fixed

File: Objects/dictobject.c (hunk 5 above)
Comment (ericsnowcurrently) [reply Objects/dictobject.c:2094]: fixed

**Review by methane — APPROVED** [review]

**Review by sunmy2019 — COMMENTED** [review]
LGTM! now it's much cleaner

**Review by sunmy2019 — APPROVED** [review]
LGTM! now it's much cleaner

**Comment by ericsnowcurrently** [comment]
Thanks for the reviews!

**Comment by eduardo-elizondo** [comment]
Nice fix! I just got around to looking at this PR, it was weird thinking about the PyDictKeysObject since they were not PyObjects, so I just let them be. That being said, this LGTM specially if we are trying to share as much state as possible.
